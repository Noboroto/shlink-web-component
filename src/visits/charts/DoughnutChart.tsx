import { isDarkThemeEnabled, PRIMARY_DARK_COLOR, PRIMARY_LIGHT_COLOR } from '@shlinkio/shlink-frontend-kit';
import type { Chart, ChartData, ChartDataset, ChartOptions } from 'chart.js';
import { keys, values } from 'ramda';
import type { FC } from 'react';
import { memo, useState } from 'react';
import { Doughnut } from 'react-chartjs-2';
import { renderPieChartLabel } from '../../utils/helpers/charts';
import type { Stats } from '../types';
import { DoughnutChartLegend } from './DoughnutChartLegend';

interface DoughnutChartProps {
  stats: Stats;
}

const generateChartDatasets = (data: number[]): ChartDataset[] => [
  {
    data,
    backgroundColor: [
      '#97BBCD',
      '#F7464A',
      '#46BFBD',
      '#FDB45C',
      '#949FB1',
      '#57A773',
      '#414066',
      '#08B2E3',
      '#B6C454',
      '#DCDCDC',
      '#463730',
    ],
    borderColor: isDarkThemeEnabled() ? PRIMARY_DARK_COLOR : PRIMARY_LIGHT_COLOR,
    borderWidth: 2,
  },
];
const generateChartData = (labels: string[], data: number[]): ChartData => ({
  labels,
  datasets: generateChartDatasets(data),
});

export const DoughnutChart: FC<DoughnutChartProps> = memo(({ stats }) => {
  // Cannot use useRef here, as we need to re-render as soon as the ref is set
  const [chartRef, setChartRef] = useState<Chart>();
  const labels = keys(stats);
  const data = values(stats);

  const options: ChartOptions = {
    plugins: {
      legend: { display: false },
      tooltip: {
        intersect: true,
        callbacks: { label: renderPieChartLabel },
      },
    },
  };
  const chartData = generateChartData(labels, data);

  return (
    <div className="row">
      <div className="col-sm-12 col-md-7">
        <Doughnut
          height={300}
          data={chartData as any}
          options={options as any}
          ref={(element) => {
            if (element) {
              setChartRef(element as Chart);
            }
          }}
        />
      </div>
      <div className="col-sm-12 col-md-5">
        {chartRef && <DoughnutChartLegend chart={chartRef} />}
      </div>
    </div>
  );
});
