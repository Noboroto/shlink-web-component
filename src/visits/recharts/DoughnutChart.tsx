import { isDarkThemeEnabled, PRIMARY_DARK_COLOR, PRIMARY_LIGHT_COLOR } from '@shlinkio/shlink-frontend-kit';
import type { FC } from 'react';
import { Fragment, useMemo } from 'react';
import { Cell, Pie, PieChart, ResponsiveContainer, Tooltip } from 'recharts';
import { prettify } from '../../utils/helpers/numbers';
import type { Stats } from '../types';
import { CHART_TOOLTIP_STYLES, COLORS } from './constants';
import { DoughnutChartLegend } from './DoughnutChartLegend';

export type DoughnutChartProps = {
  stats: Stats;
  showNumbersInLegend: boolean;

  /** Test seam. For tests, a responsive container cannot be used */
  dimensions?: { width: number; height: number };
};

export type DoughnutChartEntry = {
  name: string;
  value: number;
  color: string;
};

export const DoughnutChart: FC<DoughnutChartProps> = ({ stats, showNumbersInLegend, dimensions }) => {
  const chartData = useMemo((): DoughnutChartEntry[] => Object.entries(stats).map(([name, value], index) => (
    { name, value, color: COLORS[index % COLORS.length] }
  )), [stats]);
  const borderColor = isDarkThemeEnabled() ? PRIMARY_DARK_COLOR : PRIMARY_LIGHT_COLOR;
  const ChartWrapper = dimensions ? Fragment : ResponsiveContainer;

  return (
    <div className="row align-items-center">
      <div className="col-sm-12 col-md-7">
        <div style={dimensions ?? { width: '100%', height: 300 }}>
          <ChartWrapper>
            <PieChart {...dimensions}>
              <Tooltip formatter={prettify} contentStyle={CHART_TOOLTIP_STYLES} itemStyle={{ color: 'white' }} />
              <Pie data={chartData} dataKey="value" nameKey="name" startAngle={360} endAngle={0} outerRadius="100%" innerRadius="50%">
                {chartData.map((entry, index) => (
                  <Cell key={`${entry.name}-${index}`} fill={entry.color} stroke={borderColor} />
                ))}
              </Pie>
            </PieChart>
          </ChartWrapper>
        </div>
      </div>
      <div className="col-sm-12 col-md-5">
        <DoughnutChartLegend chartData={chartData} showNumbers={showNumbersInLegend} />
      </div>
    </div>
  );
};
