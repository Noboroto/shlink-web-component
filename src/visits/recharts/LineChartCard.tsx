import { countBy } from '@shlinkio/data-manipulation';
import {
  HIGHLIGHTED_COLOR,
  isDarkThemeEnabled,
  MAIN_COLOR,
} from '@shlinkio/shlink-frontend-kit';
import {
  add,
  differenceInDays,
  differenceInHours,
  differenceInMonths,
  differenceInWeeks,
  endOfISOWeek,
  format, parseISO,
  startOfISOWeek,
} from 'date-fns';
import type { FC } from 'react';
import { Fragment, useCallback, useMemo, useState } from 'react';
import {
  Card,
  CardBody,
  CardHeader,
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  UncontrolledDropdown,
} from 'reactstrap';
import { CartesianGrid, Line, LineChart, ResponsiveContainer, Tooltip, XAxis, YAxis } from 'recharts';
import { formatInternational } from '../../utils/dates/helpers/date';
import { rangeOf } from '../../utils/helpers';
import { prettify } from '../../utils/helpers/numbers';
import type { NormalizedVisit, Stats } from '../types';
import { CHART_TOOLTIP_STYLES } from './constants';

type ChartPayloadEntry = {
  date: string;
  amount: number;
  highlightedAmount: number;
};

const STEPS_MAP = {
  monthly: 'Month',
  weekly: 'Week',
  daily: 'Day',
  hourly: 'Hour',
} as const satisfies Record<string, string>;

type Step = keyof typeof STEPS_MAP;

const STEP_TO_DURATION_MAP: Record<Step, (amount: number) => Duration> = {
  hourly: (hours: number) => ({ hours }),
  daily: (days: number) => ({ days }),
  weekly: (weeks: number) => ({ weeks }),
  monthly: (months: number) => ({ months }),
};

const STEP_TO_DIFF_FUNC_MAP: Record<Step, (dateLeft: Date, dateRight: Date) => number> = {
  hourly: differenceInHours,
  daily: differenceInDays,
  weekly: differenceInWeeks,
  monthly: differenceInMonths,
};

const STEP_TO_DATE_FORMAT: Record<Step, (date: Date) => string> = {
  hourly: (date) => format(date, 'yyyy-MM-dd HH:00'),
  // TODO Fix formatInternational return type
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  daily: (date) => formatInternational(date)!,
  weekly(date) {
    // TODO Fix formatInternational return type
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const firstWeekDay = formatInternational(startOfISOWeek(date))!;
    // TODO Fix formatInternational return type
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const lastWeekDay = formatInternational(endOfISOWeek(date))!;

    return `${firstWeekDay} - ${lastWeekDay}`;
  },
  monthly: (date) => format(date, 'yyyy-MM'),
};

const determineInitialStep = (oldestVisitDate: string): Step => {
  const now = new Date();
  const oldestDate = parseISO(oldestVisitDate);
  const conditions: [() => boolean, Step][] = [
    [() => differenceInDays(now, oldestDate) <= 2, 'hourly'], // Less than 2 days
    [() => differenceInMonths(now, oldestDate) <= 1, 'daily'], // Between 2 days and 1 month
    [() => differenceInMonths(now, oldestDate) <= 6, 'weekly'], // Between 1 and 6 months
  ];

  return conditions.find(([matcher]) => matcher())?.[1] ?? 'monthly';
};

const countVisitsByDate = (step: Step, visits: NormalizedVisit[]): Stats => countBy(
  visits,
  (visit) => STEP_TO_DATE_FORMAT[step](parseISO(visit.date)),
);

const visitsToDatasetGroups = (step: Step, visits: NormalizedVisit[]): Record<string, NormalizedVisit[]> =>
  visits.reduce<Record<string, NormalizedVisit[]>>(
    (acc, visit) => {
      const key = STEP_TO_DATE_FORMAT[step](parseISO(visit.date));

      acc[key] = acc[key] ?? [];
      acc[key].push(visit);

      return acc;
    },
    {},
  );

const baseStatsWithNoGaps = (step: Step, visits: NormalizedVisit[]): Stats => {
  // We assume the list of visits is ordered, so the first and last visit should have the bigger and smaller dates
  const firstVisit = visits[0];
  const lastVisit = visits[visits.length - 1];

  if (!firstVisit || !lastVisit) {
    return {};
  }

  const diffFunc = STEP_TO_DIFF_FUNC_MAP[step];
  const formatter = STEP_TO_DATE_FORMAT[step];
  const duration = STEP_TO_DURATION_MAP[step];
  const newerDate = parseISO(firstVisit.date);
  const oldestDate = parseISO(lastVisit.date);
  const size = diffFunc(newerDate, oldestDate);

  const labels = [
    formatter(oldestDate),
    ...rangeOf(size, (num) => formatter(add(oldestDate, duration(num)))),
  ];

  return labels.reduce<Stats>((stats, label) => {
    // eslint-disable-next-line no-param-reassign
    stats[label] = 0;
    return stats;
  }, {});
};

type VisitsLineOptions = {
  color: string;
  dataKey: string;
  onDotClick: any;
};

// Using a function instead of an actual component because lines do not get render in that case. Need to investigate
const renderLine = ({ onDotClick, dataKey, color }: VisitsLineOptions) => (
  <Line
    type="monotone"
    dataKey={dataKey}
    stroke={color}
    strokeWidth={3}
    activeDot={{
      cursor: 'pointer',
      onClick: onDotClick,
    }}
  />
);

export type LineChartCardProps = {
  title: string;
  highlightedLabel?: string;
  visits: NormalizedVisit[];
  highlightedVisits: NormalizedVisit[];
  setSelectedVisits?: (visits: NormalizedVisit[]) => void;

  /** Test seam. For tests, a responsive container cannot be used */
  dimensions?: { width: number; height: number };
};

export const LineChartCard: FC<LineChartCardProps> = (
  { visits, title, highlightedVisits, highlightedLabel = 'Selected', setSelectedVisits, dimensions },
) => {
  const [step, setStep] = useState<Step>(
    visits.length > 0 ? determineInitialStep(visits[visits.length - 1].date) : 'monthly',
  );

  const chartData = useMemo((): ChartPayloadEntry[] => {
    const mainVisitsStats = countVisitsByDate(step, [...visits].reverse());
    const highlightedVisitsStats = countVisitsByDate(step, [...highlightedVisits].reverse());
    const baseStats = { ...baseStatsWithNoGaps(step, visits), ...mainVisitsStats };

    return Object.entries(baseStats).map(([date, amount]) => ({
      date,
      amount,
      highlightedAmount: highlightedVisitsStats[date] ?? 0,
    }));
  }, [step, visits, highlightedVisits]);

  // Save a map of step/date and all visits which belong to it, to use it when we need to highlight one
  const datasetsByPoint = useMemo(() => visitsToDatasetGroups(step, visits), [step, visits]);

  const onDotClick = useCallback((_: any, { payload }: { payload: ChartPayloadEntry }) => {
    const visitsToHighlight = datasetsByPoint[payload.date] ?? [];
    setSelectedVisits?.(visitsToHighlight === highlightedVisits ? [] : visitsToHighlight);
  }, [datasetsByPoint, highlightedVisits, setSelectedVisits]);

  const ChartWrapper = dimensions ? Fragment : ResponsiveContainer;
  const wrapperDimensions = dimensions ? {} : { width: '100%', height: 400 /* TODO Set 300 in mobile devices */ };

  return (
    <Card>
      <CardHeader role="heading" aria-level={4}>
        {title}
        <div className="float-end">
          <UncontrolledDropdown>
            <DropdownToggle caret color="link" className="btn-sm p-0">
              Group by
            </DropdownToggle>
            <DropdownMenu end>
              {Object.entries(STEPS_MAP).map(([value, menuText]) => (
                <DropdownItem key={value} active={step === value} onClick={() => setStep(value as Step)}>
                  {menuText}
                </DropdownItem>
              ))}
            </DropdownMenu>
          </UncontrolledDropdown>
        </div>
      </CardHeader>
      <CardBody>
        <ChartWrapper {...wrapperDimensions}>
          <LineChart data={chartData} {...dimensions}>
            <XAxis dataKey="date" />
            <YAxis dataKey="amount" tickFormatter={prettify} />
            <Tooltip
              formatter={(value: number, name) => [prettify(value), name === 'amount' ? 'Visits' : highlightedLabel]}
              contentStyle={CHART_TOOLTIP_STYLES}
            />
            <CartesianGrid strokeOpacity={isDarkThemeEnabled() ? 0.1 : 0.9} />
            {renderLine({ color: MAIN_COLOR, dataKey: 'amount', onDotClick })}
            {highlightedVisits.length > 0 && renderLine({ color: HIGHLIGHTED_COLOR, dataKey: 'highlightedAmount', onDotClick })}
          </LineChart>
        </ChartWrapper>
      </CardBody>
    </Card>
  );
};
