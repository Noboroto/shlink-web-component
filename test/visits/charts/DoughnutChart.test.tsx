import { screen } from '@testing-library/react';
import { DoughnutChart } from '../../../src/visits/recharts/DoughnutChart';
import { checkAccessibility } from '../../__helpers__/accessibility';
import { renderWithEvents } from '../../__helpers__/setUpTest';

describe('<DoughnutChart />', () => {
  const stats = { foo: 123, bar: 456 };
  const dimensions = { width: 800, height: 300 };
  const setUp = () => renderWithEvents(<DoughnutChart stats={stats} dimensions={dimensions} showNumbersInLegend />);

  it('passes a11y checks', () => checkAccessibility(setUp()));

  it('renders Doughnut with expected props', () => {
    const { container } = setUp();
    expect(container).toMatchSnapshot();
  });

  it('renders expected legend', () => {
    setUp();

    expect(screen.getByText('foo')).toBeInTheDocument();
    expect(screen.getByText('bar')).toBeInTheDocument();
  });
});
