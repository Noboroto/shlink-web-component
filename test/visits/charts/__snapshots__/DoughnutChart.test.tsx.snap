// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<DoughnutChart /> > renders Doughnut with expected props 1`] = `
[
  {
    "props": {
      "a": 1,
      "b": 0,
      "c": 0,
      "d": 1,
      "e": 0,
      "f": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "setTransform",
  },
]
`;
